src/kv/lazy-store.test.ts

```diff
@@ -0,0 +1,639 @@
+import {expect} from '@esm-bundle/chai';
+import {runAll} from './store-test-util';
+import {TestMemStore} from './test-mem-store';
+import {LazyStore} from './lazy-store';
+import type {Value} from './store';
+
+runAll(
+  'lazystore everything pinned',
+  () => new LazyStore(new TestMemStore(), 0, _ => true),
+);
+
+runAll(
+  'lazystore everything cached (large cache limit, nothing pinned)',
+  () => new LazyStore(new TestMemStore(), 10000, _ => false),
+);
+
+const DEFAULT_VALUE_SIZE = 100;
+function getSizeOfValueForTest(value: Value): number {
+  if (value && typeof value === 'object' && !Array.isArray(value)) {
+    const {size} = value as Record<string, Value>;
+    if (typeof size === 'number') {
+      return size;
+    }
+  }
+  return DEFAULT_VALUE_SIZE;
+}
+
+function shouldBePinnedForTest(key: string): boolean {
+  return key.startsWith('pinned');
+}
+
+const DEFAULT_CACHE_SIZE_LIMIT = 200;
+function createLazyStoreForTest(
+  options: {
+    cacheSizeLimit?: number;
+  } = {},
+) {
+  const {cacheSizeLimit = DEFAULT_CACHE_SIZE_LIMIT} = options;
+  const baseStore = new TestMemStore();
+  return {
+    baseStore,
+    lazyStore: new LazyStore(
+      baseStore,
+      cacheSizeLimit,
+      shouldBePinnedForTest,
+      getSizeOfValueForTest,
+    ),
+  };
+}
+
+test('put does not write through to base store', async () => {
+  const {baseStore, lazyStore} = createLazyStoreForTest();
+  await lazyStore.withWrite(async write => {
+    await write.put('notPinned1', 'notPinned1Value');
+    await write.commit();
+  });
+  await baseStore.withRead(async read => {
+    expect(await read.get('notPinned1')).to.be.undefined;
+  });
+});
+
+test('del does not write through to base store', async () => {
+  const {baseStore, lazyStore} = createLazyStoreForTest();
+  await baseStore.withWrite(async write => {
+    await write.put('inBase1', 'inBase1Value');
+    await write.commit();
+  });
+  await lazyStore.withWrite(async write => {
+    await write.del('inBase1');
+    await write.commit();
+  });
+  await baseStore.withRead(async read => {
+    expect(await read.get('notPinned1')).to.be.undefined;
```

Comment:
Is this test testing the right thing?

I believe it should be:

```ts
test('del does not write through to base store', async () => {
  const {baseStore, lazyStore} = createLazyStoreForTest();
  await baseStore.withWrite(async write => {
    await write.put('inBase1', 'inBase1Value');
    await write.commit();
  });
  await lazyStore.withWrite(async write => {
    await write.del('inBase1');
    await write.commit();
  });
  await lazyStore.withRead(async read => {
    expect(await read.get('inBase1')).to.be.undefined;
  });
  await baseStore.withRead(async read => {
    expect(await read.get('inBase1')).to.equal('inBase1Value');
  });
});
```

Reply:
good catch.  looks like a copy-pasta mistake.

rewrote to

```
test('del does not write through to base store', async () => {
  const {baseStore, lazyStore} = createLazyStoreForTest();
  await baseStore.withWrite(async write => {
    await write.put('inBase1', 'inBase1Value');
    await write.put('inBase2', 'inBase2Value');
    await write.commit();
  });
  await lazyStore.withRead(async read => {
    // load one of keys into the lazy store cache
    expect(await read.get('inBase1')).to.equal('inBase1Value');
  });
  await lazyStore.withWrite(async write => {
    // delete both keys (cached and not-cached) from lazy store
    await write.del('inBase1');
    await write.del('inBase2');
    await write.commit();
  });
  await baseStore.withRead(async read => {
    // neither delete impacts baseStore
    expect(await read.get('inBase1')).to.equal('inBase1Value');
    expect(await read.get('inBase1')).to.equal('inBase1Value');
  });
});
```